stan/math/gpu/matrix_gpu.hpp

```diff
@@ -0,0 +1,283 @@
+      : rows_(A.rows()), cols_(A.cols()) {
+    if (size() > 0) {
+      cl::Context& ctx = opencl_context.context();
+      cl::CommandQueue& queue = opencl_context.queue();
+      try {
+        // creates the OpenCL buffer to copy the Eigen
+        // matrix to the OpenCL device
+        oclBuffer_
+            = cl::Buffer(ctx, CL_MEM_READ_WRITE, sizeof(double) * A.size());
+        /** Writes the contents of A to the OpenCL buffer
+         * starting at the offset 0.
+         * CL_TRUE denotes that the call is blocking as
+         * we do not want to execute any further kernels
+         * on the device until we are sure that the data
+         * is finished transfering)
+         */
+        queue.enqueueWriteBuffer(oclBuffer_, CL_TRUE, 0, sizeof(T) * A.size(),
+                                 A.data());
+      } catch (const cl::Error& e) {
+        check_opencl_error("matrix constructor", e);
+      }
+    }
+  }
+
+  matrix_gpu& operator=(const matrix_gpu& a) {
+    check_size_match("assignment of GPU matrices", "source.rows()", a.rows(),
+                     "destination.rows()", rows());
+    check_size_match("assignment of GPU matrices", "source.cols()", a.cols(),
+                     "destination.cols()", cols());
+    oclBuffer_ = a.buffer();
+    return *this;
+  }
+};
+
+/**
+ * Copies the source Eigen matrix to
+ * the destination matrix that is stored
+ * on the GPU.
+ *
+ * @tparam T type of data in the Eigen matrix
+ * @param src source Eigen matrix
+ * @param dst destination matrix on the GPU
+ *
+ * @throw <code>std::invalid_argument</code> if the
+ * matrices do not have matching dimensions
+ *
+ */
+template <typename T, int R, int C>
+void copy(const Eigen::Matrix<T, R, C>& src, matrix_gpu& dst) {
+  check_size_match("copy (Eigen -> GPU)", "src.rows()", src.rows(),
+                   "dst.rows()", dst.rows());
+  check_size_match("copy (Eigen -> GPU)", "src.cols()", src.cols(),
+                   "dst.cols()", dst.cols());
+  if (src.size() > 0) {
+    cl::CommandQueue queue = opencl_context.queue();
+    try {
+      /** writes the contents of src to the OpenCL buffer
+       * starting at the offset 0
+       * CL_TRUE denotes that the call is blocking
+       * We do not want to execute any further kernels
+       * on the device until we are sure that the data is transferred)
+       */
+      queue.enqueueWriteBuffer(dst.buffer(), CL_TRUE, 0, sizeof(T) * dst.size(),
+                               src.data());
+    } catch (const cl::Error& e) {
+      check_opencl_error("copy Eigen->GPU", e);
+    }
+  }
+}
+
+/**
+ * Copies the source matrix that is stored
+ * on the GPU to the destination Eigen
+ * matrix.
+ *
+ * @tparam T type of data in the Eigen matrix
+ * @param src source matrix on the GPU
+ * @param dst destination Eigen matrix
+ *
+ * @throw <code>std::invalid_argument</code> if the
+ * matrices do not have matching dimensions
+ *
+ */
+template <typename T, int R, int C>
```

Comment:
Will `T` ever be `stan::math::var`? 
- If not, we should take away the template parameter and make that `double`. 
- If so, we should test it just to make sure it works. I saw some manual allocation with `size_of(double)` -- just verifying that it does the right thing would be good.

Reply:
> Will T ever be stan::math::var?

No idt so. We can set that to double.

Reply:
Actually. Today the answer is no. But in the future that could be a thing. 

@rok-cesnovar I think we would have to go through and template all of this, but how do you feel about passing vars onto the gpu?

Reply:
Using C++ classes in kernel code is not yet supported (it was announced a month ago and it will be included in OpenCL 2.2), so we can't handle that for now. Even if we were to switch to 2.2, we would severely limit the number of supported devices if we did that.

Reply:
Cool then let's remove the T and only accept accepting doubles (or we could limit T to just arithmetic types)